// fire/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.Worker = exports.transaction = exports.Switch = exports.style = exports.Store = exports.startWorkers = exports.startup = exports.sql = exports.setTheme = exports.Service = exports.Schema = exports.Route = exports.registerService = exports.registerInjectProvider = exports.Redirect = exports.React = exports.raw = exports.parse = exports.observer = exports.Table = exports.login = exports.Log = exports.Link = exports.isServer = exports.isClient = exports.inject = exports.fetch = exports.deny = exports.createServer = exports.configureWebpack = exports.component = exports.classNames = exports.check = exports.allow = exports.addRegisterProvider = exports.addInjectProvider = exports.addShims = undefined;

var _fireLog = require('fire-log');

var _fireLog2 = _interopRequireDefault(_fireLog);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _typeCheckSystem = require('type-check-system');

var _typeCheckSystem2 = _interopRequireDefault(_typeCheckSystem);

var _nodeFetch = require('node-fetch');

var _nodeFetch2 = _interopRequireDefault(_nodeFetch);

var _fireWebpack = require('fire-webpack');

var _legoSql = require('lego-sql');

var _legoSql2 = _interopRequireDefault(_legoSql);

var _reactRouterDom = require('react-router-dom');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _sqlModels = require('sql-models');

var _index = require('./component/index.js');

var _index2 = _interopRequireDefault(_index);

var _index3 = require('./service/index.js');

var _index4 = require('./store/index.js');

var _index5 = _interopRequireDefault(_index4);

var _index6 = require('./injector/index.js');

var _index7 = require('./server/index.js');

var _index8 = _interopRequireDefault(_index7);

var _serverService = require('./service/server-service.js');

var _index9 = require('./style/index.js');

var _index10 = require('./worker/index.js');

var _index11 = _interopRequireDefault(_index10);

var _index12 = require('./startup/index.js');

var _index13 = _interopRequireDefault(_index12);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var sql = _legoSql2.default.sql;
var transaction = _legoSql2.default.transaction;
var parse = _legoSql2.default.parse;
var raw = _legoSql2.default.raw;

function isRunning() {
    return !process.env.FIRE_STAGE || process.env.FIRE_STAGE === 'start';
}

function isClient() {
    return isRunning() && typeof window !== 'undefined';
}

function isServer() {
    return isRunning() && !isClient();
}

var registerInjectProvider = function registerInjectProvider() {
    // TODO: This is deprecated.

    return _index6.addInjectProvider.apply(undefined, arguments);
};

exports.addShims = _fireWebpack.addShims;
exports.addInjectProvider = _index6.addInjectProvider;
exports.addRegisterProvider = _index6.addRegisterProvider;
exports.allow = _serverService.allow;
exports.check = _typeCheckSystem2.default;
exports.classNames = _classnames2.default;
exports.component = _index2.default;
exports.configureWebpack = _fireWebpack.configureWebpack;
exports.createServer = _index8.default;
exports.deny = _serverService.deny;
exports.fetch = _nodeFetch2.default;
exports.inject = _index6.inject;
exports.isClient = isClient;
exports.isServer = isServer;
exports.Link = _reactRouterDom.Link;
exports.Log = _fireLog2.default;
exports.login = _serverService.login;
exports.Table = _sqlModels.Table;
exports.observer = _index.observer;
exports.parse = parse;
exports.raw = raw;
exports.React = _react2.default;
exports.Redirect = _reactRouterDom.Redirect;
exports.registerInjectProvider = registerInjectProvider;
exports.registerService = _index3.registerService;
exports.Route = _reactRouterDom.Route;
exports.Schema = _sqlModels.Schema;
exports.Service = _index3.Service;
exports.setTheme = _index9.setTheme;
exports.sql = sql;
exports.startup = _index13.default;
exports.startWorkers = _index10.startWorkers;
exports.Store = _index5.default;
exports.style = _index9.style;
exports.Switch = _reactRouterDom.Switch;
exports.transaction = transaction;
exports.Worker = _index11.default;
